### PR TITLE
[FLINK-21417][core] Separate type-specific memory segments.

### PR DESCRIPTION
**Pull request by xintongsong**
## What is the purpose of the change

*(For example: This pull request makes task deployment go through the blob server, rather than through RPC. That way we avoid re-transferring them on each deployment (during recovery).)*

## Brief change log

*(for example:)*
  - *The TaskInfo is stored in the blob store on job creation time as a persistent artifact*
  - *Deployments RPC transmits only the blob storage reference*
  - *TaskManagers retrieve the TaskInfo from the blob cache*

## Verifying this change

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (100MB)*
  - *Extended integration test for recovery after master (JobManager) failure*
  - *Added test that validates that TaskInfo is transferred only once across recoveries*
  - *Manually verified the change by running a 4 node cluser with 2 JobManagers and 4 TaskManagers, a stateful streaming program, and killing one JobManager and two TaskManagers during the execution, verifying that recovery happens correctly.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / no)
  - The serializers: (yes / no / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / no / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / no / don't know)
  - The S3 file system connector: (yes / no / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by xintongsong** [comment]
cc @tillrohrmann

File: flink-core/src/main/java/org/apache/flink/core/memory/MemorySegmentFactory.java [hunk 1]
```diff
@@ -31,7 +31,7 @@
 import static org.apache.flink.util.Preconditions.checkArgument;
 
 /**
- * A factory for (hybrid) memory segments ({@link HybridMemorySegment}).
+ * A factory for (hybrid) memory segments ({@link OffHeapMemorySegment}).
```
Comment (tillrohrmann) [inline flink-core/src/main/java/org/apache/flink/core/memory/MemorySegmentFactory.java:34]: `A factory for {@link OffHeapMemorySegment}`.

File: flink-core/src/main/java/org/apache/flink/core/memory/OffHeapMemorySegment.java [hunk 2]
```diff
@@ -24,33 +24,22 @@
...
- * <p>The memory can be on-heap, off-heap direct or off-heap unsafe, this is transparently handled
- * by this class.
- *
- * <p>This class specializes byte access and byte copy calls for heap memory, while reusing the
- * multi-byte type accesses and cross-segment operations from the MemorySegment.
- *
- * <p>This class subsumes the functionality of the {@link
- * org.apache.flink.core.memory.HeapMemorySegment}, but is a bit less efficient for operations on
- * individual bytes.
+ * <p>The memory can direct or unsafe, this is transparently handled by this class.
```
Comment (tillrohrmann) [inline flink-core/src/main/java/org/apache/flink/core/memory/OffHeapMemorySegment.java:33]: `can be`

File: flink-core/src/test/java/org/apache/flink/core/memory/HybridOffHeapDirectMemorySegmentTest.java [hunk 3]
```diff
@@ -51,7 +51,7 @@ MemorySegment createSegment(int size, Object owner) {
     @Test
     public void testHybridHeapSegmentSpecifics() {
         final int bufSize = 411;
-        HybridMemorySegment seg = (HybridMemorySegment) createSegment(bufSize);
+        OffHeapMemorySegment seg = (OffHeapMemorySegment) createSegment(bufSize);
```
Comment (tillrohrmann) [inline flink-core/src/test/java/org/apache/flink/core/memory/HybridOffHeapDirectMemorySegmentTest.java:54]: The cast should not be necessary if we change the return type of `createSegment`.

File: flink-core/src/test/java/org/apache/flink/core/memory/HybridOffHeapDirectMemorySegmentTest.java [hunk 4]
```diff
@@ -51,7 +51,7 @@ MemorySegment createSegment(int size, Object owner) {
     @Test
     public void testHybridHeapSegmentSpecifics() {
```
Comment (tillrohrmann) [inline flink-core/src/test/java/org/apache/flink/core/memory/HybridOffHeapDirectMemorySegmentTest.java:52]: Name of the test can be updated.

File: flink-core/src/test/java/org/apache/flink/core/memory/HybridOffHeapDirectMemorySegmentTest.java [hunk 5]
```diff
@@ -30,7 +30,7 @@
 import static org.junit.Assert.assertTrue;
 import static org.junit.Assert.fail;
 
-/** Tests for the {@link HybridMemorySegment} in off-heap mode using direct memory. */
+/** Tests for the {@link OffHeapMemorySegment} in off-heap mode using direct memory. */
 @RunWith(Parameterized.class)
 public class HybridOffHeapDirectMemorySegmentTest extends MemorySegmentTestBase {
```
Comment (tillrohrmann) [inline flink-core/src/test/java/org/apache/flink/core/memory/HybridOffHeapDirectMemorySegmentTest.java]: I guess this test will be renamed in a later commit, right?

File: flink-core/src/test/java/org/apache/flink/core/memory/HybridOffHeapUnsafeMemorySegmentTest.java [hunk 6]
```diff
@@ -26,7 +26,7 @@
 
 import static org.junit.Assert.assertTrue;
 
-/** Tests for the {@link HybridMemorySegment} in off-heap mode using unsafe memory. */
+/** Tests for the {@link OffHeapMemorySegment} in off-heap mode using unsafe memory. */
 @RunWith(Parameterized.class)
 public class HybridOffHeapUnsafeMemorySegmentTest extends MemorySegmentTestBase {
```
Comment (tillrohrmann) [inline flink-core/src/test/java/org/apache/flink/core/memory/HybridOffHeapUnsafeMemorySegmentTest.java]: Will be renamed later, right?

File: flink-core/src/main/java/org/apache/flink/core/memory/MemorySegmentFactory.java [hunk 7]
```diff
@@ -130,7 +130,7 @@ public static MemorySegment allocateUnpooledOffHeapMemory(int size) {
      */
     public static MemorySegment allocateUnpooledOffHeapMemory(int size, Object owner) {
```
Comment (tillrohrmann) [inline flink-core/src/main/java/org/apache/flink/core/memory/MemorySegmentFactory.java]: Should we rename the factory methods to reflect which type of off heap memory they allocate?

File: flink-core/src/main/java/org/apache/flink/core/memory/UnsafeMemorySegment.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+     *
+     * <p>The memory segment references the given owner.
+     *
+     * @param buffer The byte buffer whose memory is represented by this memory segment.
+     * @param owner The owner references by this memory segment.
+     * @param cleaner The cleaner to be called on free segment.
+     * @throws IllegalArgumentException Thrown, if the given ByteBuffer is not direct.
+     */
+    UnsafeMemorySegment(
+            @Nonnull ByteBuffer buffer, @Nullable Object owner, @Nullable Runnable cleaner) {
```
Comment (tillrohrmann) [inline flink-core/src/main/java/org/apache/flink/core/memory/UnsafeMemorySegment.java:52]: When is `cleaner` `null`?

File: flink-core/src/test/java/org/apache/flink/core/memory/EndiannessAccessChecks.java [hunk 9]
```diff
@@ -37,19 +37,18 @@ public void testHeapSegment() {
 
     @Test
     public void testHybridOnHeapSegment() {
-        testBigAndLittleEndianAccessUnaligned(MemorySegmentFactory.wrap(new byte[11111]));
+        testBigAndLittleEndianAccessUnaligned(
+                MemorySegmentFactory.wrapHeapSegment(new byte[11111]));
     }
 
     @Test
     public void testHybridOffHeapSegment() {
```
Comment (tillrohrmann) [inline flink-core/src/test/java/org/apache/flink/core/memory/EndiannessAccessChecks.java:45]: could be renamed into `testDirectOffHeapSegment`

File: flink-core/src/test/java/org/apache/flink/core/memory/EndiannessAccessChecks.java [hunk 10]
```diff
@@ -37,19 +37,18 @@ public void testHeapSegment() {
...
-        testBigAndLittleEndianAccessUnaligned(
-                MemorySegmentFactory.allocateUnpooledOffHeapMemory(11111));
+        testBigAndLittleEndianAccessUnaligned(MemorySegmentFactory.allocateDirectSegment(11111));
     }
 
     @Test
     public void testHybridOffHeapUnsafeSegment() {
```
Comment (tillrohrmann) [inline flink-core/src/test/java/org/apache/flink/core/memory/EndiannessAccessChecks.java:50]: Could be renamed into `testUnsafeOffHeapSegment`

**Review by tillrohrmann — APPROVED** [review]
Nice work @xintongsong. I think the changes look good to me. I had a few comments which we should resolve before merging this PR.

File: flink-core/src/main/java/org/apache/flink/core/memory/MemorySegmentFactory.java (hunk 7 above)
Comment (xintongsong) [reply flink-core/src/main/java/org/apache/flink/core/memory/MemorySegmentFactory.java]: Addressed in later commit.

File: flink-core/src/main/java/org/apache/flink/core/memory/UnsafeMemorySegment.java (hunk 8 above)
Comment (xintongsong) [reply flink-core/src/main/java/org/apache/flink/core/memory/UnsafeMemorySegment.java:52]: Nice catch. We should no longer have null cleaner.

**Comment by xintongsong** [comment]
Thanks for the quick response and review, @tillrohrmann.
Comments addressed. I'll merge this once AZP gives green light.

**Pull request closed without merging**